reflex/event.py

```diff
@@ -1047,32 +1055,157 @@ def call_event_handler(
     Args:
         event_handler: The event handler.
         arg_spec: The lambda that define the argument(s) to pass to the event handler.
+        key: The key to pass to the event handler.
 
     Raises:
         EventHandlerArgMismatch: if number of arguments expected by event_handler doesn't match the spec.
+        EventHandlerArgTypeMismatch: if the annotations of args accepted by event_handler differs from the spec of the event trigger.
+        TypeError: if a failure occurs while comparing types.
 
     Returns:
         The event spec from calling the event handler.
+
+    # noqa: DAR401 failure
+
     """
     parsed_args = parse_args_spec(arg_spec)  # type: ignore
 
     if isinstance(event_handler, EventSpec):
         # Handle partial application of EventSpec args
         return event_handler.add_args(*parsed_args)
 
-    args = inspect.getfullargspec(event_handler.fn).args
-    n_args = len(args) - 1  # subtract 1 for bound self arg
-    if n_args == len(parsed_args):
-        return event_handler(*parsed_args)  # type: ignore
-    else:
+    provided_callback_fullspec = inspect.getfullargspec(event_handler.fn)
+
+    provided_callback_n_args = (
+        len(provided_callback_fullspec.args) - 1
+    )  # subtract 1 for bound self arg
+
+    if provided_callback_n_args != len(parsed_args):
         raise EventHandlerArgMismatch(
             "The number of arguments accepted by "
-            f"{event_handler.fn.__qualname__} ({n_args}) "
+            f"{event_handler.fn.__qualname__} ({provided_callback_n_args}) "
             "does not match the arguments passed by the event trigger: "
             f"{[str(v) for v in parsed_args]}\n"
             "See https://reflex.dev/docs/events/event-arguments/"
         )
 
+    def compare_types(provided_type, accepted_type):
```

Comment:
> can we move this to `reflex/utils/types.py`

yes

> i can use it for State.__setattr__ type checking

it's not built for this, as setattr will get a *value* and you want to compare that to a *type hint*, this function is supposed to compare two *type hints* to each other, which makes this function not well defined to the role here, i can find many bugs in it already !

Reply:
basically i want to `compare_type(type(v), self._get_type_hints()[name])`

Reply:
i'm not sure that would do it, because `type(v)` doesn't output a type hint, but rather the type itself, and this errs on the side of being strict, so we might get more false negatives out of that

i can possibly move it out and provide very intensive unit tests so to try to catch all possible cases, but i'm worried we're giving an even more power to a function that isn't correct and *can't be* correct